### PR TITLE
midi interface revamp

### LINKED ISSUE #223: Just updated from master repo, doesn't compile
OS: Win7
Version (or "dev" if compiling from source): Dev

src/core/MidiInterface.cpp:4:10: fatal error: rtmidi/RtMidi.h: No such file or d   irectory
 #include "rtmidi/RtMidi.h"
          ^~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [compile.mk:51: build/src/core/MidiInterface.cpp.o] Error 1

### LINKED ISSUE #238: Crash when selecting same MIDI interface
OS (if applicable): Windows 7
Version (or "dev" if compiling from source): 0.4.0
Steps to reproduce:
  - Add Midi to CV module and select a MIDI interface.
  - Add CC to CV module and select the same MIDI interface as previous step.
  - Crash

### LINKED ISSUE #239: Crash when selecting MIDI Interface in MIDI to CV module under Jack in Linux
OS (if applicable): Linux
Version (or "dev" if compiling from source): dev

To reproduce:  Start jack on Linux, run vcvrack, add MIDI to CV module, select MIDI hardware interface and vcvrack crashes.

Crash does not occur when running under ALSA.

### PR DESCRIPTION
**Pull request by bontric**
In this PR I address the following issues:

### Midi learn [DONE]
* Add midi learn to Trigger clock module [DONE]

Note: To use midi learn, right click on the cc number/trigger note text field and turn knob/press key

### Restructure[DONE]
As discussed in #201

### MIDI Trigger [DONE]
As discussed in #228 
* 16 outputs which correspond to the customizable midi numbers. ~~Each ouput ouputs a value in relation to the velocity.~~ [DONE]
### MIDI Clock interface [DONE]
I used this Molten [CV-SYNC](http://www.moltenvoltage.com/products/CV_SYNC_by_Molten_Voltage_-_MIDI_Clock_to_CV_Synchronizer.html) Module as reference. 

However, we could easily add more ratios. 

TODO: 
+ ~~If you switch the ratio, you need to start/stop the external clock~~ [FIXED]
+ ~~If you change the clock's speed you need to start/stop the external clock~~ [FIXED]
+ ~~Add cv input for ratio selection.~~[DONE]
+ ~~Add "Continue" message output~~ (I think a reset out is good enough for the clock)

(See also #201)

### CODE REVIEW

File: src/core/MidiInterface.hpp [hunk 1]
```diff
@@ -13,25 +13,25 @@ using namespace rack;
  */
 
 struct MidiInWrapper : RtMidiIn {
-	std::unordered_map<uint, std::list<std::vector<unsigned char>>> idMessagesMap;
-	std::unordered_map<uint, std::list<double>> idStampsMap;
-	uint uuid_c = 0;
-	uint subscribers = 0;
```
Comment (bontric) [inline src/core/MidiInterface.hpp]: Note: (Regarding `uint`). It's not defined by default by all compilers, see [here](https://stackoverflow.com/questions/3552094/c-uint-unsigned-int-int)

**Comment by bontric** [comment]
@AndrewBelt aside from some small improvements, this is almost ready for merge. Would be great if you could take a look and give some feedback.

**Comment by bontric** [comment]
@AndrewBelt could you add a  "virtual onTextSelected"  function to TextField?  
This would allow me to easily implement midi Learn for CC/Trigger module. I had the idea and it works almost out of the box when I overwrite 
```c
void CCTextField::draw(NVGcontext *vg) {
	/* This is necessary, since the save
	 * file is loaded after constructing the widget*/
	if (*inited) {
		*inited = false;
		text = std::to_string(*ccNum);
	}

	if (*onFocus) {
		text = std::to_string(*ccNum);
	}

	TextField::draw(vg);
}

bool CCTextField::onFocusText(int codepoint) {
	*onFocus = false;
	return TextField::onFocusText(codepoint);
}

Widget *CCTextField::onMouseDown(Vec pos, int button) {
	*onFocus = true;
	return OpaqueWidget::onMouseDown(pos, button);
}
```

And then change the CC number value if a knob is turned while text is selected:

```c		
	if (status == 0xb) {
		for (int i = 0; i < NUM_OUTPUTS; i++) {
			if (onFocus[i]) {
				this->ccNum[i] = data1;
			}
		}
		for (int i = 0; i < NUM_OUTPUTS; i++) {
			if (data1 == ccNum[i]) {
				this->cc[i] = data2;
			}
		}
	}
```

Btw.. I need to say: The code base is great. It seems like everything I want to do is just a few lines of code without any hassle.

**Comment by bontric** [comment]
@AndrewBelt This should be ready for merge

**Comment by AndrewBelt** [comment]
Fantastic! I'll add a couple ratios, mainly on the fast side.

A couple things I may add: note names in trigger-to-CV text boxes. I want to type in C2 and assign that to a kick. But MIDI learn is better anyway. Just thought it should be added.

The rest looks great. No complaints. Looking ready for v0.5.0. I'll get to work on the Audio Interface migration to RtAudio and then take off from there. Thanks!

**Pull request merged**